src/libData/AccountData/EvmClient.cpp

```diff
@@ -0,0 +1,132 @@
+#include <boost/range/iterator_range.hpp>
+#include <thread>
+
+#include "EvmClient.h"
+#include "libUtils/DetachedFunction.h"
+#include "libUtils/EvmJsonResponse.h"
+#include "libUtils/EvmUtils.h"
+#include "libUtils/JsonUtils.h"
+#include "libUtils/SysCommand.h"
+
+/* EvmClient Init */
+void EvmClient::Init() {
+  LOG_MARKER();
+
+  CheckClient(0, false);
+
+  m_initialised = true;
+}
+
+bool EvmClient::OpenServer(bool force) {
+  LOG_MARKER();
+
+  std::string programName =
+      boost::filesystem::path(EVM_SERVER_BINARY).filename().string();
+  std::string cmdStr = "pkill " + programName + " ; " + EVM_SERVER_BINARY +
+                       " --socket " + EVM_SERVER_SOCKET_PATH +
+                       " --tracing >/dev/null &";
+
+  LOG_GENERAL(INFO, "running cmdStr: " << cmdStr);
+
+  try {
+    if (!SysCommand::ExecuteCmd(SysCommand::WITHOUT_OUTPUT, cmdStr)) {
+      LOG_GENERAL(WARNING, "ExecuteCmd failed: " << cmdStr);
+      return false;
+    }
+  } catch (const std::exception& e) {
+    LOG_GENERAL(WARNING,
+                "Exception caught in SysCommand::ExecuteCmd: " << e.what());
+    return false;
+  } catch (...) {
+    LOG_GENERAL(WARNING, "Unknown error encountered");
+    return false;
+  }
+
+  LOG_GENERAL(WARNING, "Executed: " << cmdStr);
+
+  std::this_thread::sleep_for(
+      std::chrono::milliseconds(SCILLA_SERVER_PENDING_IN_MS));
+
+  if (force) {
+    m_initialised = false;
+    CheckClient(0, false);
+    m_initialised = true;
+  }
+
+  return true;
+}
+
+bool EvmClient::CheckClient(uint32_t version, bool enforce) {
+  std::lock_guard<std::mutex> g(m_mutexMain);
+
+  if (m_initialised) return true;
+
+  if (!OpenServer(enforce)) {
+    LOG_GENERAL(WARNING, "OpenServer for version " << version << "failed");
+    return false;
+  }
+
+  std::shared_ptr<jsonrpc::UnixDomainSocketClient> conn =
+      std::make_shared<jsonrpc::UnixDomainSocketClient>(EVM_SERVER_SOCKET_PATH);
+
+  m_connectors[version] = conn;
+
+  std::shared_ptr<jsonrpc::Client> c = std::make_shared<jsonrpc::Client>(
+      *m_connectors.at(version), jsonrpc::JSONRPC_CLIENT_V2);
+  m_clients[version] = c;
+
+  return true;
+}
+
+bool EvmClient::CallRunner(uint32_t version, const Json::Value& _json,
+                           evmproj::CallRespose& result, uint32_t counter) {
+  if (counter == 0) {
+    return false;
+  }
+
+  version = 0;
+
+  if (!CheckClient(version)) {
+    LOG_GENERAL(WARNING, "CheckClient failed");
+    return false;
+  }
+
+  try {
+    std::lock_guard<std::mutex> g(m_mutexMain);
+    Json::Value oldJson;
+    evmproj::CallRespose reply;
+    oldJson = m_clients.at(version)->CallMethod("run", _json);
+    // Populate the C++ struct with the return values
+    reply = evmproj::GetReturn(oldJson, result);
+  } catch (jsonrpc::JsonRpcException& e) {
+    LOG_GENERAL(WARNING, "CallRunner failed: " << e.what());
+    m_initialised = false;
+    if (!CheckClient(version, true)) {
+      LOG_GENERAL(WARNING,
```

Comment:
I think at the minium the warning message is wrong here. But please reconsider error handling, why do we need CheckClient after run fails?

Reply:
The exception handling has been tightened up in this version, RPC calling errors are distinguished from Json Parsing errors that are all caught individually reported and rethrown to this handler.

This Handler now reports the error as :

 LOG_GENERAL(WARNING, "RPC Exception calling EVM-DS : attempting server "
                "restart " << e.what());
                
This is handling the error in a similar fashion to legacy code but is more respectful in its restarting of the server, legacy restarts each time.

This code will shortly be superceded with gRPC.